src/rpc/blockchain.cpp

```diff
@@ -518,6 +518,69 @@ UniValue getblockbynumber(const UniValue& params, bool fHelp)
     return blockToJSON(block, pblockindex, params.size() > 1 ? params[1].get_bool() : false);
 }
 
+UniValue getblocksbatch(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 3)
+        throw runtime_error(
+                "getblocksbatch <starting number> <number of blocks> [bool:txinfo]\n"
+                "\n"
+                "<starting number> the block number at the start of the batch"
+                "\n"
+                "<number of blocks> the number of blocks to return in the batch, limited to 500"
+                "\n"
+                "[bool:txinfo] optional to print more detailed tx info\n"
+                "\n"
+                "Returns a JSON array with details of the requested blocks starting with\n"
+                "the given block-number.\n");
+
+    UniValue result(UniValue::VOBJ);
+    UniValue blocks(UniValue::VARR);
+
+    int nHeight = params[0].get_int();
+    if (nHeight < 0 || nHeight > nBestHeight) throw runtime_error("Starting block number out of range");
+
+    int batch_size = params[1].get_int();
+    if (batch_size < 1 || batch_size > 500)
```

Comment:
Why limit the batch_size to 500? The wallet will be frozen for a while but anyone who uses this RPC is probably aware of that.

Reply:
Precisely for the reason that you just said. An API should always be built with thoughtful limits to keep people out of trouble, not built with the assumption that they will be used properly. Without a reasonable limit, someone could put in there getblocksbatch 0 5000000 true and try to get the whole block chain in one go. That is completely non-sensical.

There are reasonable limits to how much JSON to be parsed on the caller side, and also how long we should be holding the cs_main lock. When that lock is held, nothing else can be done in the core.

If the calling environment is multithreaded, it will cause other simultaneous rpcs to pile up while this lock is being held.

The number is somewhat arbitrary. 500 is also the interval batch size used for block transmission for syncing.

Since this will be called most likely by an external loop, you simply need a high enough batch size to render the rpc calling overhead to be very small compared to the rpc transaction. 1/500 is a pretty small overhead.